Python 3.9 support

Hey, I am getting the following errors in the test suite on Python 3.9. Is this expected? I have looked at the code but it is not very clear to me what is failing here, if you give me some pointer I may be able to fix it.

```
============================= test session starts ==============================
platform linux -- Python 3.9.0, pytest-6.1.2, py-1.9.0, pluggy-0.13.1
rootdir: /build/python-flake8-typing-imports/src/flake8-typing-imports-1.10.0
collected 26 items

tests/flake8_typing_imports_test.py ............FFFFFx........           [100%]

=================================== FAILURES ===================================
_____________________ test_union_pattern_or_match[Pattern] _____________________

s = 'from typing import Pattern, Union\ndef foo(bar: Union[Pattern, str]): pass\n'

    @pytest.mark.parametrize(
        's', (
            pytest.param(
                'from typing import Pattern, Union\n'
                'def foo(bar: Union[Pattern, str]): pass\n',
                id='Pattern',
            ),
            pytest.param(
                'import typing\n'
                'def foo(bar: typing.Union[typing.Pattern, str]): pass\n',
                id='typing.Pattern',
            ),
            pytest.param(
                'from typing import Match, Union\n'
                'def foo(bar: Union[Match, str]): pass\n',
                id='Match',
            ),
            pytest.param(
                'import typing\n'
                'def foo(bar: typing.Union[typing.Match, str]): pass\n',
                id='typing.Match',
            ),
            pytest.param(
                'from typing import Match, Pattern, Union\n'
                'def foo(bar: Union[Match, Pattern, int]): pass\n',
                id='Match and Pattern',
            ),
            pytest.param(
                'from typing import Match as M, Union\n'
                'def foo(bar: Union[M, str]): pass\n',
                id='Match imported as Name',
                marks=pytest.mark.xfail(
                    reason=(
                        'this is broken too, but so unlikely we elect not to '
                        'detect it'
                    ),
                ),
            ),
        ),
    )
    def test_union_pattern_or_match(s):
        with version_ctx(Version(3, 5, 0)):
>           assert results(s) == {
                '2:13: TYP003 Union[Match, ...] or Union[Pattern, ...] '
                'must be quoted in <3.5.2',
            }
E           AssertionError: assert set() == {'2:13: TYP00...ed in <3.5.2'}
E             Extra items in the right set:
E             '2:13: TYP003 Union[Match, ...] or Union[Pattern, ...] must be quoted in <3.5.2'
E             Use -v to get the full diff

/build/python-flake8-typing-imports/src/flake8-typing-imports-1.10.0/tests/flake8_typing_imports_test.py:170: AssertionError
_________________ test_union_pattern_or_match[typing.Pattern] __________________

s = 'import typing\ndef foo(bar: typing.Union[typing.Pattern, str]): pass\n'

    @pytest.mark.parametrize(
        's', (
            pytest.param(
                'from typing import Pattern, Union\n'
                'def foo(bar: Union[Pattern, str]): pass\n',
                id='Pattern',
            ),
            pytest.param(
                'import typing\n'
                'def foo(bar: typing.Union[typing.Pattern, str]): pass\n',
                id='typing.Pattern',
            ),
            pytest.param(
                'from typing import Match, Union\n'
                'def foo(bar: Union[Match, str]): pass\n',
                id='Match',
            ),
            pytest.param(
                'import typing\n'
                'def foo(bar: typing.Union[typing.Match, str]): pass\n',
                id='typing.Match',
            ),
            pytest.param(
                'from typing import Match, Pattern, Union\n'
                'def foo(bar: Union[Match, Pattern, int]): pass\n',
                id='Match and Pattern',
            ),
            pytest.param(
                'from typing import Match as M, Union\n'
                'def foo(bar: Union[M, str]): pass\n',
                id='Match imported as Name',
                marks=pytest.mark.xfail(
                    reason=(
                        'this is broken too, but so unlikely we elect not to '
                        'detect it'
                    ),
                ),
            ),
        ),
    )
    def test_union_pattern_or_match(s):
        with version_ctx(Version(3, 5, 0)):
>           assert results(s) == {
                '2:13: TYP003 Union[Match, ...] or Union[Pattern, ...] '
                'must be quoted in <3.5.2',
            }
E           AssertionError: assert set() == {'2:13: TYP00...ed in <3.5.2'}
E             Extra items in the right set:
E             '2:13: TYP003 Union[Match, ...] or Union[Pattern, ...] must be quoted in <3.5.2'
E             Use -v to get the full diff

/build/python-flake8-typing-imports/src/flake8-typing-imports-1.10.0/tests/flake8_typing_imports_test.py:170: AssertionError
______________________ test_union_pattern_or_match[Match] ______________________

s = 'from typing import Match, Union\ndef foo(bar: Union[Match, str]): pass\n'

    @pytest.mark.parametrize(
        's', (
            pytest.param(
                'from typing import Pattern, Union\n'
                'def foo(bar: Union[Pattern, str]): pass\n',
                id='Pattern',
            ),
            pytest.param(
                'import typing\n'
                'def foo(bar: typing.Union[typing.Pattern, str]): pass\n',
                id='typing.Pattern',
            ),
            pytest.param(
                'from typing import Match, Union\n'
                'def foo(bar: Union[Match, str]): pass\n',
                id='Match',
            ),
            pytest.param(
                'import typing\n'
                'def foo(bar: typing.Union[typing.Match, str]): pass\n',
                id='typing.Match',
            ),
            pytest.param(
                'from typing import Match, Pattern, Union\n'
                'def foo(bar: Union[Match, Pattern, int]): pass\n',
                id='Match and Pattern',
            ),
            pytest.param(
                'from typing import Match as M, Union\n'
                'def foo(bar: Union[M, str]): pass\n',
                id='Match imported as Name',
                marks=pytest.mark.xfail(
                    reason=(
                        'this is broken too, but so unlikely we elect not to '
                        'detect it'
                    ),
                ),
            ),
        ),
    )
    def test_union_pattern_or_match(s):
        with version_ctx(Version(3, 5, 0)):
>           assert results(s) == {
                '2:13: TYP003 Union[Match, ...] or Union[Pattern, ...] '
                'must be quoted in <3.5.2',
            }
E           AssertionError: assert set() == {'2:13: TYP00...ed in <3.5.2'}
E             Extra items in the right set:
E             '2:13: TYP003 Union[Match, ...] or Union[Pattern, ...] must be quoted in <3.5.2'
E             Use -v to get the full diff

/build/python-flake8-typing-imports/src/flake8-typing-imports-1.10.0/tests/flake8_typing_imports_test.py:170: AssertionError
__________________ test_union_pattern_or_match[typing.Match] ___________________

s = 'import typing\ndef foo(bar: typing.Union[typing.Match, str]): pass\n'

    @pytest.mark.parametrize(
        's', (
            pytest.param(
                'from typing import Pattern, Union\n'
                'def foo(bar: Union[Pattern, str]): pass\n',
                id='Pattern',
            ),
            pytest.param(
                'import typing\n'
                'def foo(bar: typing.Union[typing.Pattern, str]): pass\n',
                id='typing.Pattern',
            ),
            pytest.param(
                'from typing import Match, Union\n'
                'def foo(bar: Union[Match, str]): pass\n',
                id='Match',
            ),
            pytest.param(
                'import typing\n'
                'def foo(bar: typing.Union[typing.Match, str]): pass\n',
                id='typing.Match',
            ),
            pytest.param(
                'from typing import Match, Pattern, Union\n'
                'def foo(bar: Union[Match, Pattern, int]): pass\n',
                id='Match and Pattern',
            ),
            pytest.param(
                'from typing import Match as M, Union\n'
                'def foo(bar: Union[M, str]): pass\n',
                id='Match imported as Name',
                marks=pytest.mark.xfail(
                    reason=(
                        'this is broken too, but so unlikely we elect not to '
                        'detect it'
                    ),
                ),
            ),
        ),
    )
    def test_union_pattern_or_match(s):
        with version_ctx(Version(3, 5, 0)):
>           assert results(s) == {
                '2:13: TYP003 Union[Match, ...] or Union[Pattern, ...] '
                'must be quoted in <3.5.2',
            }
E           AssertionError: assert set() == {'2:13: TYP00...ed in <3.5.2'}
E             Extra items in the right set:
E             '2:13: TYP003 Union[Match, ...] or Union[Pattern, ...] must be quoted in <3.5.2'
E             Use -v to get the full diff

/build/python-flake8-typing-imports/src/flake8-typing-imports-1.10.0/tests/flake8_typing_imports_test.py:170: AssertionError
________________ test_union_pattern_or_match[Match and Pattern] ________________

s = 'from typing import Match, Pattern, Union\ndef foo(bar: Union[Match, Pattern, int]): pass\n'

    @pytest.mark.parametrize(
        's', (
            pytest.param(
                'from typing import Pattern, Union\n'
                'def foo(bar: Union[Pattern, str]): pass\n',
                id='Pattern',
            ),
            pytest.param(
                'import typing\n'
                'def foo(bar: typing.Union[typing.Pattern, str]): pass\n',
                id='typing.Pattern',
            ),
            pytest.param(
                'from typing import Match, Union\n'
                'def foo(bar: Union[Match, str]): pass\n',
                id='Match',
            ),
            pytest.param(
                'import typing\n'
                'def foo(bar: typing.Union[typing.Match, str]): pass\n',
                id='typing.Match',
            ),
            pytest.param(
                'from typing import Match, Pattern, Union\n'
                'def foo(bar: Union[Match, Pattern, int]): pass\n',
                id='Match and Pattern',
            ),
            pytest.param(
                'from typing import Match as M, Union\n'
                'def foo(bar: Union[M, str]): pass\n',
                id='Match imported as Name',
                marks=pytest.mark.xfail(
                    reason=(
                        'this is broken too, but so unlikely we elect not to '
                        'detect it'
                    ),
                ),
            ),
        ),
    )
    def test_union_pattern_or_match(s):
        with version_ctx(Version(3, 5, 0)):
>           assert results(s) == {
                '2:13: TYP003 Union[Match, ...] or Union[Pattern, ...] '
                'must be quoted in <3.5.2',
            }
E           AssertionError: assert set() == {'2:13: TYP00...ed in <3.5.2'}
E             Extra items in the right set:
E             '2:13: TYP003 Union[Match, ...] or Union[Pattern, ...] must be quoted in <3.5.2'
E             Use -v to get the full diff

/build/python-flake8-typing-imports/src/flake8-typing-imports-1.10.0/tests/flake8_typing_imports_test.py:170: AssertionError
=========================== short test summary info ============================
FAILED tests/flake8_typing_imports_test.py::test_union_pattern_or_match[Pattern]
FAILED tests/flake8_typing_imports_test.py::test_union_pattern_or_match[typing.Pattern]
FAILED tests/flake8_typing_imports_test.py::test_union_pattern_or_match[Match]
FAILED tests/flake8_typing_imports_test.py::test_union_pattern_or_match[typing.Match]
FAILED tests/flake8_typing_imports_test.py::test_union_pattern_or_match[Match and Pattern]
=================== 5 failed, 20 passed, 1 xfailed in 0.16s ====================
```

--- a/azure-pipelines.yml
+++ b/azure-pipelines.yml
@@ -10,10 +10,10 @@ resources:
       type: github
       endpoint: github
       name: asottile/azure-pipeline-templates
-      ref: refs/tags/v1.0.0
+      ref: refs/tags/v2.0.0
 
 jobs:
 - template: job--python-tox.yml@asottile
   parameters:
-    toxenvs: [py36, py37]
+    toxenvs: [py36, py37, py38, py39]
     os: linux

--- a/flake8_typing_imports.py
+++ b/flake8_typing_imports.py
@@ -18,6 +18,17 @@ else:  # pragma: no cover (PY38+)
     import importlib.metadata as importlib_metadata
 
 
+if sys.version_info < (3, 9):  # pragma: no cover (<PY39)
+    def _get_subscript_slice(node: ast.Subscript) -> ast.AST:
+        if isinstance(node.slice, ast.Index):
+            return node.slice.value
+        else:
+            return node.slice
+else:  # pragma: no cover (PY39+)
+    def _get_subscript_slice(node: ast.Subscript) -> ast.AST:
+        return node.slice
+
+
 class Version(NamedTuple):
     major: int = 0
     minor: int = 0
@@ -856,14 +867,14 @@ class Visitor(ast.NodeVisitor):
         self.generic_visit(node)
 
     def visit_Subscript(self, node: ast.Subscript) -> None:
+        slice_value = _get_subscript_slice(node)
         if (
             self._is_typing(node.value, ('Union',)) and
-            isinstance(node.slice, ast.Index) and
-            isinstance(node.slice.value, ast.Tuple) and
-            len(node.slice.value.elts) > 1 and
+            isinstance(slice_value, ast.Tuple) and
+            len(slice_value.elts) > 1 and
             any(
                 self._is_typing(x, ('Pattern', 'Match'))
-                for x in node.slice.value.elts
+                for x in slice_value.elts
             )
         ):
             self.unions_pattern_or_match.append((node.lineno, node.col_offset))

--- a/tox.ini
+++ b/tox.ini
@@ -1,5 +1,5 @@
 [tox]
-envlist = py36,py37,py38,pre-commit
+envlist = py36,py37,py38,py39,pre-commit
 
 [testenv]
 deps = -rrequirements-dev.txt

